### PR TITLE
rpc: feature flags, macro changes, misc setup

### PR DESCRIPTION
**Pull request by hinto-janai**
### What
This PR:
- Makes changes to the base RPC types
- Adds a documentation macro the whole crate can use
- Adds the `default` and `free` (internal) module for default/conditional serde behavior
- Makes changes to the type generator macro
    - More maintenance friendly (sub-macros instead of duplicated branches)
    - Supports default field values
- Adds the `misc` module containing _every single_ misc type used in other request/response types

---

This is the 1st out of 4 PRs that implement types in `cuprate-rpc-types`.

There are 3 PRs built on-top of this one:
1. https://github.com/Cuprate/cuprate/pull/219
2. https://github.com/Cuprate/cuprate/pull/220
3. https://github.com/Cuprate/cuprate/pull/221

These respectively implement all:
- `/json_rpc` types
- `/*.bin` types
- Other JSON types

### CODE REVIEW

File: rpc/types/Cargo.toml [hunk 1]
```diff
@@ -9,14 +9,16 @@ repository  = "https://github.com/Cuprate/cuprate/tree/main/rpc/types"
 keywords    = ["cuprate", "rpc", "types", "monero"]
 
 [features]
-default = []
+default = ["serde", "epee"]
+serde   = ["dep:serde"]
+epee    = ["dep:cuprate-epee-encoding"]
```
Comment (hinto-janai) [inline rpc/types/Cargo.toml:14]: All types enabled by default, only `serde` and `epee` are feature gated (enabled by default though).

File: rpc/types/README.md [hunk 2]
```diff
@@ -10,13 +10,14 @@ This crate ports the types used in Monero's RPC interface, including:
...
-- [`other`] module; Misc JSON types from other endpoints
-
-Miscellaneous types are found in the root module, e.g. [`crate::Status`].
+| Module | Purpose |
+|--------|---------|
+| The root module | Miscellaneous items, e.g. constants.
+| [`json`] | Contains JSON request/response (some mixed with binary) that all share the common `/json_rpc` endpoint. |
+| [`bin`] | Contains request/response types that are expected to be fully in binary (`cuprate_epee_encoding`) in `monerod` and `cuprated`'s RPC interface. These are called at a custom endpoint instead of `/json_rpc`, e.g. `/get_blocks.bin`. |
+| [`other`] | Contains request/response types that are JSON, but aren't called at `/json_rpc` (e.g. [`crate::other::GetHeightRequest`]). |
+| [`misc`] | Contains miscellaneous types, e.g. [`crate::misc::Status`]. Many of types here are found and used in request/response types, for example, [`crate::misc::BlockHeader`] is used in [`crate::json::GetLastBlockHeaderResponse`]. |
```
Comment (hinto-janai) [inline rpc/types/README.md:19]: All misc types are now in the `misc` module instead of root.

File: rpc/types/src/base.rs [hunk 3]
```diff
@@ -10,76 +10,30 @@
 //! - <https://github.com/monero-project/monero/commit/2899379791b7542e4eb920b5d9d58cf232806937>
 //! - <https://github.com/monero-project/monero/issues/8722>
 //! - <https://github.com/monero-project/monero/pull/8843>
+//!
+//! This library doesn't contain [`rpc_access_request_base`](https://github.com/monero-project/monero/blob/cc73fe71162d564ffda8e549b79a350bca53c454/src/rpc/core_rpc_server_commands_defs.h#L114-L122)
+//! found in `monerod` as the type is practically deprecated.
+//!
+//! Although, [`AccessResponseBase`] still exists as to allow
+//! outputting the same JSON fields as `monerod` (even if deprecated).
```
Comment (hinto-janai) [inline rpc/types/src/base.rs:18]: There is a `rpc_access_request_base` defined in `monerod` that pairs with the `rpc_access_response_base`.

This contains a `std::string: client` field, but since this is deprecated, I removed the implementation as we'll never use it - however - I kept `AccessResponseBase` as this actually affects the JSON output (i.e. it adds `top_hash: String`, `credits: u64`).

At the end of the day, the response is deprecated too but I thought following `monerod`'s JSON output more than not would be good. We can also remove this if wanted, they're always dead fields anyway.

File: rpc/types/src/base.rs [hunk 4]
```diff
@@ -10,76 +10,30 @@
...
-    ($start:literal..=$end:literal) => {
-        concat!(
-            "[Definition](https://github.com/monero-project/monero/blob/cc73fe71162d564ffda8e549b79a350bca53c454/src/rpc/core_rpc_server_commands_defs.h#L",
-            stringify!($start),
-            "-L",
-            stringify!($end),
-            ")."
-        )
-    };
-}
```
Comment (hinto-janai) [inline rpc/types/src/base.rs:33]: Moved to `macros.rs` for the whole crate to use.

File: rpc/types/src/defaults.rs [hunk 5]
```diff
@@ -0,0 +1,78 @@
+//! These functions define the default values
+//! of optional fields in request/response types.
+//!
+//! For example, [`crate::json::GetBlockRequest`]
+//! has a [`crate::json::GetBlockRequest::height`]
+//! field and a [`crate::json::GetBlockRequest::hash`]
+//! field, when the RPC interface reads JSON without
+//! `height`, it will use [`default_height`] to fill that in.
```
Comment (hinto-janai) [inline rpc/types/src/defaults.rs:8]: These functions exist to pass to `serde` and `epee` for default field values.

Our own free functions are used over something like `Default::default` because:
- It's more explicit
- Sometimes we need the non-default, e.g. `default_bool_true`.

There's no example in this PR, but these are extensively used in the other PRs.

File: rpc/types/src/free.rs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+//! Free functions.
+
+//---------------------------------------------------------------------------------------------------- Serde
+// These are functions used for conditionally (de)serialization.
```
Comment (hinto-janai) [inline rpc/types/src/free.rs:4]: Same as `defaults.rs`, this allows for conditional (de)serialization in some cases.

Examples are in the other PRs.

File: rpc/types/src/json.rs [hunk 7]
```diff
@@ -3,10 +3,7 @@
...
 
 //---------------------------------------------------------------------------------------------------- Import
-use crate::{
-    base::{EmptyRequestBase, EmptyResponseBase, ResponseBase},
-    macros::define_request_and_response,
-};
+use crate::{base::ResponseBase, macros::define_request_and_response};
 
 //---------------------------------------------------------------------------------------------------- Struct definitions
 // This generates 2 structs:
```
Comment (hinto-janai) [inline rpc/types/src/json.rs:9]: Any `{json,bin,other.rs}` changes can be ignored in this PR, they get implemented in the other PRs.

File: rpc/types/src/lib.rs [hunk 8]
```diff
@@ -94,23 +102,25 @@
     )
 )]
 // TODO: remove me after finishing impl
-#![allow(dead_code)]
+#![allow(
+	dead_code,
+    rustdoc::broken_intra_doc_links // TODO: remove after `{bin,json,other}.rs` gets merged
```
Comment (hinto-janai) [inline rpc/types/src/lib.rs:107]: TODO: keep this unresolved until done.

File: rpc/types/src/macros.rs [hunk 9]
```diff
@@ -17,35 +15,35 @@
...
+/// flatten that into the request type automatically.
 ///
-/// FIXME: there's probably a less painful way to branch here on input
-/// without having to duplicate 80% of the macro. Sub-macros were attempted
-/// but they ended up unreadable. So for now, make sure to fix the other
-/// branch as well when making changes. The only de-duplicated part is
-/// the doc generation with [`define_request_and_response_doc`].
+/// E.g. `Response {/*...*/}` and `ResponseBase {/*...*/}`
+/// would trigger the different branches.
 macro_rules! define_request_and_response {
```
Comment (hinto-janai) [inline rpc/types/src/macros.rs:46]: The type generator macro no longer has duplicated code, it only has 1 branch and uses sub-macros instead.

Since we got rid of the RPC access `Request` stuff, this macro _always_ takes `Request {}` as input now. No fields will create the `type Request = ()`, where fields will create a normal request.

Responses still use the RPC access syntax, so it's one of:
- `Request` - adds nothing (some requests only have `status`, so we need this)
- `ResponseBase` - adds `base: crate::base::ResponseBase`
- `AccessResponseBase` - adds `base: crate::base::AccessResponseBase`

File: rpc/types/src/macros.rs [hunk 10]
```diff
@@ -59,177 +57,234 @@ macro_rules! define_request_and_response {
...
+        $( #[$type_attr:meta] )*
         $type_name:ident,
 
+        // The request type (and any doc comments, derives, etc).
+        $( #[$request_type_attr:meta] )*
+        Request {
+            // And any fields.
+            $(
+                $( #[$request_field_attr:meta] )*
+                $request_field:ident: $request_field_type:ty $(= $request_field_type_default:expr)?,
```
Comment (hinto-janai) [inline rpc/types/src/macros.rs:71]: And also, it supports `cuprate_epee_encoding`'s default value syntax.

We can find a way to automatically do this with `serde` too (right now the caller must add the serde attribute) but that's for another PR.

File: rpc/types/src/macros.rs [hunk 11]
```diff
@@ -269,9 +323,34 @@ macro_rules! define_request_and_response_doc {
...
+            "#L",
+            stringify!($start),
+            $(
+                "-L",
+                stringify!($end),
+            )?
             ")."
         )
     };
 }
```
Comment (hinto-janai) [inline rpc/types/src/macros.rs:355]: This documents the original definition for pretty much every type in `cuprate-rpc-types`.

File: rpc/types/src/misc/binary_string.rs [hunk 12]
```diff
@@ -1,14 +1,14 @@
-//! TODO
+//! JSON string containing binary data.
 
 //---------------------------------------------------------------------------------------------------- Import
 
 //---------------------------------------------------------------------------------------------------- BinaryString
-/// TODO
+/// TODO: we need to figure out a type that (de)serializes correctly, `String` errors with `serde_json`
```
Comment (hinto-janai) [inline rpc/types/src/misc/binary_string.rs:6]: For another PR.

File: rpc/types/src/misc/block_complete_entry.rs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct BlockCompleteEntry {
+    pub pruned: bool,
+    pub block: String,
+    pub block_weight: u64,
+    pub txs: Vec<TxBlobEntry>,
+}
+
+// TODO: custom epee
+// <https://github.com/monero-project/monero/blob/cc73fe71162d564ffda8e549b79a350bca53c454/src/cryptonote_protocol/cryptonote_protocol_defs.h#L138-L163>
```
Comment (hinto-janai) [inline rpc/types/src/misc/block_complete_entry.rs:29]: Custom (de)serialization will be done in another PR.

File: rpc/types/src/misc/misc.rs [hunk 14]
```diff
@@ -0,0 +1,539 @@
+//! Miscellaneous types.
+//!
+//! These are `struct`s that appear in request/response types.
+//! For example, [`crate::json::GetConnectionsResponse`] contains
+//! the [`crate::misc::ConnectionInfo`] struct defined here.
```
Comment (hinto-janai) [inline rpc/types/src/misc/misc.rs:5]: TODO: I have to document this but `misc.rs` contains all misc types that can be created in a macro-friendly way.

The other files in the `misc/` folder are because they have something special, so the macro doesn't work.

**Review by hinto-janai — COMMENTED** [review]
Things left that will be done in another PR after this + all type impl PRs are merged:
- Custom (de)serialization
- `Status` fixes (doesn't catch all strings, probably need `Other(String)`
- Field type optimizations
- Macro optimizations (less duplicated input, etc)

File: rpc/types/src/misc/misc.rs [hunk 15]
```diff
@@ -0,0 +1,539 @@
...
+    #[doc = monero_definition_link!(
+        cc73fe71162d564ffda8e549b79a350bca53c454,
+        "rpc/core_rpc_server_commands_defs.h",
+        538..=553
+    )]
+    #[derive(Copy)]
+    /// Used in [`crate::bin::GetOutsRequest`].
+    OutKeyBin {
+        key: u8, // TODO: crypto::public_key,
+        mask: u8, // TODO: rct::key,
```
Comment (Boog900) [inline rpc/types/src/misc/misc.rs:395]: These are `[u8; 32]`

File: rpc/types/src/misc/tx_entry.rs [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    pub prunable_as_hex: String,
+    pub prunable_hash: String,
+    pub pruned_as_hex: String,
+    pub received_timestamp: u64,
+    pub relayed: bool,
+    pub tx_hash: String,
+}
+
+// TODO: custom epee
+// <https://github.com/monero-project/monero/blob/cc73fe71162d564ffda8e549b79a350bca53c454/src/cryptonote_protocol/cryptonote_protocol_defs.h#L406-427>
```
Comment (Boog900) [inline rpc/types/src/misc/tx_entry.rs:41]: Is this the correct link?

**Review by Boog900 — CHANGES_REQUESTED** [review]

File: rpc/types/src/misc/misc.rs (hunk 15 above)
Comment (hinto-janai) [reply rpc/types/src/misc/misc.rs:395]: These were going to be replace with some dedicated types and/or aliases as there's a few more of these (e.g. [`typedef std::string blobdata`](https://github.com/monero-project/monero/blob/cc73fe71162d564ffda8e549b79a350bca53c454/src/cryptonote_basic/blobdatatype.h#L39)), but maybe the type directly is okay?

File: rpc/types/src/misc/tx_entry.rs (hunk 16 above)
Comment (hinto-janai) [reply rpc/types/src/misc/tx_entry.rs]: Oops, fixed to correct link (https://github.com/monero-project/monero/blob/cc73fe71162d564ffda8e549b79a350bca53c454/src/rpc/core_rpc_server_commands_defs.h#L406-L428).

File: rpc/types/src/base.rs (hunk 3 above)
Comment (hinto-janai) [reply rpc/types/src/base.rs]: Added back `AccessRequestBase`.

FYI the empty bases (`Request`, `Response`) are special-cased by the macro since flattening an empty struct causes some errors.

**Review by Boog900 — APPROVED** [review]
